### ISSUE TITLE
Bug in Poisson link function

### ISSUE THREAD

**Issue by tyarkoni** [bambinos/bambi#111]
This was [pointed out on Stack Overflow](https://stackoverflow.com/questions/45525631/incorrect-results-with-poisson-regression-using-bambi) a while ago, but I only just saw it (we need to get better about tracking SO posts--I just created an alert, and suggest you do too, @jake-westfall). Bambi is currently giving incorrect results for Poisson regression. This turns out to be due to an incorrect link function--it's currently specified in the PyMC3 back-end as log, when it should be exp. The problem actually stems from the same issue in PyMC3, as discussed [here](https://github.com/pymc-devs/pymc3/issues/1100) (though this is entirely my fault--I should have inspected the link functions more closely when porting them over).

This is a pretty big red flag that we need to be much more comprehensive about testing results for all major model classes for reasonable convergence with other packages (e.g, statsmodels). We should probably also start testing the Stan/PyMC3 backend outputs for convergence (we currently test priors, but I don't think we test parameter estimates). That would have helped here, because the Stan back-end produces the right result. I'll try to implement something more systematic when I have time (e.g., a simple step would be to write a helper function that takes two Stan and PyMC3-fitted models and tests that all parameter estimates are reasonably close).

Fix coming momentarily; I'm just detailing the bug here for posterity.

**Comment by jake-westfall** [comment]
I agree that the function to be applied should be `theano.tensor.exp`. But basically this points to a terminological problem that we have, since `exp` is the _inverse_ link function and not the link function. So basically we are asking users for the "link function" but then applying it as if it's the inverse link function. This terminological issue is almost certainly how we got confused and ended up using `log` instead of `exp` as the default for Poisson.

Looking at our other links, "logit" is implemented correctly, since its value is `theano.tensor.nnet.sigmoid` which is the inverse of the logit function. And "inverse" is also correct since 1/x is its own inverse.

We should think about how we want to handle this. The easiest thing to do is to just have the `links` dictionary in the pymc backend contain the key:value `'log':theano.tensor.exp`. We could then probably delete the `'exp'` key in the dictionary, since AFAIK `exp` is never used as a link function -- but on the other hand there doesn't seem to be any harm in keeping it, even if it's never used in practice, and the appropriate value for that key would be `theano.tensor.log`.

A complication arises here if users want to supply their own link function, since again we are actually using this as the inverse link function (i.e., applying it to the right-hand-side of the regression equation). So probably the best, clearest thing to do is to have a separate parameter, `inverse_link`, where users can pass in custom inverse link functions. And our existing `link` parameter will only accept string arguments and will internally point to the inverse function of whatever string is supplied (as we currently do with `logit`). 

I lean toward the second option. Thoughts?

**Comment by junpenglao** [comment]
> a simple step would be to write a helper function that takes two Stan and PyMC3-fitted models and tests that all parameter estimates are reasonably close

That would be awesome and we are very interested in that as well. I have a quick look at the tests seems there is no test for backend currently?

**Comment by jake-westfall** [comment]
I'm going to start implementing my second suggestion from above (i.e., having separate `link` and `inverse_link` parameters for `Model.add()` and `.fit()`) in the `fix-poisson-link` branch with the open PR unless you object, @tyarkoni.

**Comment by tyarkoni** [comment]
Sounds good--sorry I hadn't gotten back to this yet (I just added the other issues b/c they came up as I was revising the paper).

Your solution sounds pretty good to me, but an alternative we should consider is allowing both `link` and `inverse_link` to be passed in as either a string or a callable, but changing the JSON config so that the existing Poisson and Bernoulli families are defined using the `inverse_link` rather than the `link` (which would be set, respectively, to exp and sigmoid). Then we don't have to worry about raising an exception if the user tries to pass a callable to `link`. And it's not like anyone should care whether the internal default spec uses `link` or `inverse_link`. Would that work?

@junpenglao many of the tests do fit the model with both backends, so we at least have a bunch of smoke tests. But you're right that we don't have unit tests for the backends at the moment (or for that matter, for the `Results` object). That should ideally be remedied ASAP, but I don't think either @jake-westfall or I have the time to do it, so probably I should open issues for this and hope for contributions.

I'll try to find time for the helper I mentioned above though, as that would at least take us from smoke tests to acceptance tests without very much effort. :)

**Comment by tyarkoni** [comment]
Oh, on second thought, I guess we can't invert arbitrary link functions, so we can't allow `inverse_link` to be a callable.

I guess now I'm wondering whether we should have `link` at all. Since the specification is ultimately always going to be via the inverse link in our case, maybe it's best to just bite the bullet and force the argument to be `inverse_link`. Then trying to pass `link` will raise an error. (Or we could do the opposite, and go with your first suggestion, assuming the inverse everywhere.)

Any idea how other packages like brms handle this issue? @paul-buerkner, any input would be appreciated, as you've probably run into this already...

**Comment by jake-westfall** [comment]
I'm assuming that _"so we can't allow `inverse_link` to be a callable"_ should read _"so we can't allow `link` to be a callable"_?

As for removing the `link` parameter altogether. Pretty much every GLM(M) package that I've seen asks users to supply the link function rather than the inverse link function, and AFAIK most don't even allow you to directly specify the inverse link (other than by specifying the link). There'd be nothing wrong or incoherent about asking our users to supply the inverse link instead of the link, but I'd guess that this departure from standard practice would likely cause confusion among a non-trivial minority of our users. Having two separate parameters would further clutter our API a bit, but it _might_ be worth it to avoid confusion.

Another option that we haven't discussed is to simply drop the ability to use arbitrary link functions altogether. In which case we'd just focus instead on making sure that all the links that are used 99.9% of the time in practice are supported internally. Then we could basically keep our current API. To be honest, I've never seen, read, or heard about anyone using a link function other than the ones mentioned in [this Wikipedia table](https://en.wikipedia.org/wiki/Generalized_linear_model#Link_function). So my guess is that if we just added the one link function from that table that we're currently missing, that'd pretty much cover the overwhelming majority of use cases. Then users who want to use more exotic link functions would have to go directly to Stan/PyMC3. (In fact there's a good chance that if they need this functionality, their model already has other features that put it outside of the scope of `bambi` anyway.)

**Comment by tyarkoni** [comment]
+1 to restricting `link` to just predefined strings.

**Comment by jake-westfall** [comment]
Okay, and any thoughts about whether we want to add the `inverse_link` parameter (accepting only callable arguments) to support arbitrary link functions? I guess it's a trade-off between slightly more functionality vs. slightly simpler API (and, you know, implementation and testing time).

**Comment by tyarkoni** [comment]
Probably not worth it; I think just sticking with predefined `link` options is fine. I think my reasoning when supporting arbitrary callables passed to `link` originally was just "why not", since it was easy to do; it's not like I had any expectation that it would be used often (if ever).

**Comment by junpenglao** [comment]
FYI, just come across this which you might also find interesting:
In tensorflow, calling `link` function is actually `inverse_link` as well:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distributions/python/ops/estimator.py#L146-L151
